src/suites/cts/validation/validation_test.ts

```diff
@@ -1,16 +1,103 @@
+import { unreachable } from '../../../framework/index.js';
+import { bindingTypeInfo } from '../format_info.js';
 import { GPUTest } from '../gpu_test.js';
 
+export enum BindingResourceType {
+  'uniform-buffer' = 'uniform-buffer',
+  'storage-buffer' = 'storage-buffer',
+  'sampler' = 'sampler',
+  'sampled-texture' = 'sampled-texture',
+  'storage-texture' = 'storage-texture',
+}
+
 export class ValidationTest extends GPUTest {
-  async getErrorBuffer(): Promise<GPUBuffer> {
+  getStorageBuffer(): GPUBuffer {
+    return this.device.createBuffer({ size: 1024, usage: GPUBufferUsage.STORAGE });
+  }
+
+  getUniformBuffer(): GPUBuffer {
+    return this.device.createBuffer({ size: 1024, usage: GPUBufferUsage.UNIFORM });
+  }
+
+  getErrorBuffer(): GPUBuffer {
     this.device.pushErrorScope('validation');
     const errorBuffer = this.device.createBuffer({
       size: 1024,
       usage: 0xffff, // Invalid GPUBufferUsage
     });
-    await this.device.popErrorScope();
+    this.device.popErrorScope();
     return errorBuffer;
   }
 
+  getSampler(): GPUSampler {
+    return this.device.createSampler();
+  }
+
+  getErrorSampler(): GPUSampler {
+    this.device.pushErrorScope('validation');
+    const sampler = this.device.createSampler({ lodMinClamp: -1 });
+    this.device.popErrorScope();
+    return sampler;
+  }
+
+  getSampledTexture(): GPUTexture {
+    return this.device.createTexture({
+      size: { width: 16, height: 16, depth: 1 },
+      format: 'rgba8unorm',
+      usage: GPUTextureUsage.SAMPLED,
+    });
+  }
+
+  getStorageTexture(): GPUTexture {
+    return this.device.createTexture({
+      size: { width: 16, height: 16, depth: 1 },
+      format: 'rgba8unorm',
+      usage: GPUTextureUsage.STORAGE,
+    });
+  }
+
+  getErrorTextureView(): GPUTextureView {
+    this.device.pushErrorScope('validation');
+    const view = this.device
+      .createTexture({
+        size: { width: 0, height: 0, depth: 0 },
+        format: 'rgba8unorm',
+        usage: GPUTextureUsage.SAMPLED,
+      })
+      .createView();
+    this.device.popErrorScope();
+    return view;
+  }
+
+  getBindingResource(bindingType: BindingResourceType, error: boolean = false): GPUBindingResource {
+    if (error) {
+      const objectType = bindingTypeInfo[bindingType].type;
+      if (objectType === 'buffer') {
+        return { buffer: this.getErrorBuffer() };
+      } else if (objectType === 'sampler') {
+        return this.getErrorSampler();
+      } else if (objectType === 'texture') {
+        return this.getErrorTextureView();
+      } else {
+        unreachable('unknown binding resource type');
+      }
+    } else {
+      if (bindingType === 'uniform-buffer') {
```

Comment:
Can these use the const enums?

Reply:
Yes, and have to if I do the next suggestion

Reply:
I guess I won't do that one, so I won't do this one for the same reason. These are already typo guarded:

```
This condition will always return 'false' since the types 'BindingResourceType.storage-texture' and '"storage-bexture"' have no overlap.ts(2367)
```